arroyo/backends/kafka/consumer.py

```diff
@@ -246,40 +252,74 @@ def assignment_callback(
         ) -> None:
             self.__state = KafkaConsumerState.ASSIGNING
 
-            try:
-                assignment: MutableSequence[ConfluentTopicPartition] = []
-
-                for partition in self.__consumer.committed(partitions):
-                    if partition.offset >= 0:
-                        assignment.append(partition)
-                    elif partition.offset == OFFSET_INVALID:
-                        assignment.append(
-                            self.__resolve_partition_starting_offset(partition)
-                        )
-                    else:
-                        raise ValueError("received unexpected offset")
-
-                offsets: MutableMapping[Partition, int] = {
-                    Partition(Topic(i.topic), i.partition): i.offset for i in assignment
-                }
-                self.__seek(offsets)
-
-                # Ensure that all partitions are resumed on assignment to avoid
-                # carrying over state from a previous assignment.
-                self.__consumer.resume(
-                    [
-                        ConfluentTopicPartition(
-                            partition.topic.name, partition.index, offset
-                        )
-                        for partition, offset in offsets.items()
-                    ]
-                )
+            if self.__incremental_cooperative is True:
+                try:
+                    incremental_assignment: MutableSequence[
+                        ConfluentTopicPartition
+                    ] = []
+
+                    for partition in partitions:
```

Comment:
If I am getting this right, you are not asking the broker anymore for the last committed offset. 
Which means that you are invoking `incremental_assign` with the offset you received in the `assignment_callback` call. 
Is this intentional? In the stop-the-world rebalancing, fetching the last committed offset and seeking that was meant to ensure all partitions had the same behavior (whether they were previously assigned to the same consumer or not). All partitions would be reset to the last committed offset.
I would expect it to be impossible, in the incremental case, to receive a partition you were already owning before (unless the consumer restarted). Is this guaranteed or can you still be assigned a partition you already owned before rebalancing?
If it is guaranteed and all partitions you received are "new" to this consumer, are you not resetting the offset because the offset you receive are supposed to already be the last committed ones ?

Reply:
I don't *think* you can receive a partition you already owned, but in any case I don't quite get why we would need to check the offset since the offset you are receiving should already be the correct one. The other issue I noticed with `committed()` is that it hangs forever (or times out if you pass it a timeout) for the cooperative-sticky strategy in the scenario where it's a new consumer group and no offset has been committed yet.

Reply:
This will be important to verify. 
The reason for getting the `committed()` offsets is so that all partitions start in a consistent way from the last committed offset after a rebalance (without doing so the partitions that are reassigned to you would start from the last consumed offset). 
This behavior is visible to the application, but if we commit during revoke we should not have any issue as the two behaviors would be identical. 
If we never receive a partition we already owned and we consistently commit on revoke we are good. But if we fail in committing during revoke and we get the same partition we owned in the past I don't know what would happen if we do not reset the offsets. 
If `committed()` is taking much longer, then it is probably a non starter. Could you please verify that even if we did receive the same partition again on subscriptions (where we are testing this), we will not risk missing messages ?

Reply:
> The reason for getting the `committed()` offsets is so that all partitions start in a consistent way from the last committed offset after a rebalance (without doing so the partitions that are reassigned to you would start from the last consumed offset).

I don't think this is true. The offsets passed to the assignment callback are already the committed ones not just the ones that are consumed.

Kafka guarantees that
- only new partitions are passed in this callback not previously owned ones
- The revoke callback is always triggered before the assignment one.

The reason for committing offsets during revoke is to avoid double processing them so the new consumer doesn't get that same offset again.

Reply:
> The offsets passed to the assignment callback are already the committed ones not just the ones that are consumed.

Are you sure. I thought this was the reason for making this consumer behave differently from the standard Kafka consumer https://github.com/getsentry/arroyo/blob/main/arroyo/backends/kafka/consumer.py#L109-L118.

Anyway, if Kafka guarantees that only new partitions are passed and not previously owned one, we should be good.

Reply:
Yes, but this only applies to the partitions that are being newly assigned to the consumer. Previously all of the partitions (including those already assigned to the consumer) would be provided in the callback, now only the incremental ones are. So we are no longer rewinding back those partitions which the consumer is maintaining across a rebalance like we use to, those just continue from the same place.